### PR TITLE
Full-blown typing of VM / native compilation.

### PR DESCRIPTION
**Pull request by ppedrot**
Bracing for impact. I repeat. Bracing for impact.

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

File: tactics/redexpr.ml [hunk 1]
```diff
@@ -25,7 +25,8 @@ open Libobject
 (* call by value normalisation function using the virtual machine *)
 let cbv_vm env sigma c =
   if Coq_config.bytecode_compiler then
-    let ctyp = Retyping.get_type_of env sigma c in
+    (* Hack: drop the evarmap *)
```
Comment (SkySkimmer) [inline tactics/redexpr.ml:28]: why?

Comment (ppedrot) [reply tactics/redexpr.ml:28]: Too annoying to return it, I would have to tweak the contextual functions and whatnot. I just wanted a quick idea of the level of desolation triggered by this nuclear patch.

File: tactics/redexpr.ml [hunk 2]
```diff
@@ -25,7 +25,8 @@ open Libobject
...
+    let _sigma, ctyp = Typing.type_of env sigma c in
```
Comment (SkySkimmer) [inline tactics/redexpr.ml:29]: This is insufficient, the named_context also needs checking.
also Typing doesn't check inside evar instances, but vm/native do reduce in the evar instances

Comment (ppedrot) [reply tactics/redexpr.ml:29]: Honestly if we can't even trust the context we can just remove vm_compute.

Comment (SkySkimmer) [reply tactics/redexpr.ml:29]: also don't forget to solve remaining constraints, it's probably possible to hide a type error in them

Comment (SkySkimmer) [reply tactics/redexpr.ml:29]: Of course we can't trust the context, we can put stuff in it with `pose (x := ltac:(exact_no_check ...):...)`

Comment (andres-erbsen) [reply tactics/redexpr.ml:29]: Yep, that's [wellknown: ltac variable bypasses typechecking](https://github.com/coq/coq/issues?q=label%3A%22wellknown%3A+ltac+variable+bypasses+typechecking%22+). I think the most blatant way to exploit it now might be `Ltac2.Constr.in_context`.

Comment (SkySkimmer) [reply tactics/redexpr.ml:29]: >Yep, that's wellknown: ltac variable bypasses typechecking. 

No, that's yet another way to get ill typed terms.

pose is a regular variable not a ltac variable, typechecking is bypassed using no_check in my example.

**Comment by JasonGross** [comment]
This does not seem so bad if we gate it with a command-line flag.

**Comment by andres-erbsen** [comment]
We'll need to see what happens when the contexts are also (re)typechecked, but yes, these numbers look a lot less concerning than one might have feared.

> This does not seem so bad if we gate it with a command-line flag.

I figure a flag would be feasible, but do we really want another one? Looking at the top slowdowns, I think all the real ones are in your rewriter, and all the ones out of these that are backed by use cases are in fiat-crypto. I don't think we'd want to turn on a unsafe-to-the-point-of-memory-corruption flag in fiat-crypto, and we sure haven't taken any care to only call `vm_compute` on terms that could be typechecked in reasonable time. Looking further at individual lines:

1) The worst recorded slowown is <4s out of 12s in a [stress test](coq-engine-bench-lite/coq/PerformanceDemos/quadratic_reduction.v.html) for the vm on large inputs. Yes, we expect this would get slower. As the the large inputs are themselves generated by `vm_compute`, I think a trusted `abstract_vm_compute` could be used to avoid double-checking types here if anyone actually cared about a similar use case.
2) fiat-crypto partial evaluation of [p224](https://coq.gitlab.io/-/coq/-/jobs/3375061712/artifacts/_bench/html/coq-fiat-crypto-with-bedrock/src/Bedrock/Field/Synthesis/Examples/p224_64_new.v.html) has a nontrivial slowdown, but I wouldn't turn on a memory-corruption-threatening flag over this. Also, the overhead on Defined is much smaller than at the actual `vm_compute` call, suggesting that cleaning up the input expression could speed it up (`_no_check` is not used).
3) The Bedrock2 ethernet driver proof is either noise or should be fixed downstream -- there is no reason it should need fast compute. The [time](https://coq.gitlab.io/-/coq/-/jobs/3375061712/artifacts/_bench/html/coq-fiat-crypto-with-bedrock/rupicola/bedrock2/bedrock2/src/bedrock2Examples/lightbulb.v.html) is in Defined, which could be explained away as typechecking of `vm_cast`s left by `ring` in unnecessarily messy goals, but I don't quite believe it based on this alone or care enough to investigate.
4) fiat-crypto bootstrapping, [IdentifiersGENERATEDProofs](https://coq.gitlab.io/-/coq/-/jobs/3375061712/artifacts/_bench/html/coq-fiat-crypto-with-bedrock/src/Language/IdentifiersGENERATEDProofs.v.html) Not sure about this one. 1.5s slowdown before Qed and then Qed with barely any slowdown.
5) The [risc-v instruction decoding proof](coq-fiat-crypto-with-bedrock/rupicola/bedrock2/deps/riscv-coq/src/riscv/Proofs/DecodeByExtension.v.html) does not use vm or native based on my expectations and inspection; I expect it's spurious.

My wish would be to figure out a sound fix first, release it, and only consider potentially adding a flag once an (arguably) realistic context where good performance cannot be achieved otherwise comes up. If a flag is deemed necessary for diagnosing potential slowdowns in a point release, I wish that it would be scheduled for extermination in the next full release where it has time to go through beta and etc.

**Comment by JasonGross** [comment]
> 3\. The [time](https://coq.gitlab.io/-/coq/-/jobs/3375061712/artifacts/_bench/html/coq-fiat-crypto-with-bedrock/rupicola/bedrock2/bedrock2/src/bedrock2Examples/lightbulb.v.html) is in Defined, which could be explained away as typechecking of `vm_cast`s left by `ring` in unnecessarily messy goals, but I don't quite believe it based on this alone or care enough to investigate.

I expect that this will be fixed by a real implementation of this patch: there's no reason anymore for the kernel to typecheck the terms under a VM cast if the VM itself is going to do that (and plausibly the kernel can in some places pass an argument to the VM saying "I've checked the context already, you don't have to"), so I expect there does not need to be any slowdown in Qed/Defined.

This suggests that the only real slowdowns will be places that cannot use `abstract vm_cast_no_check`, i.e., places where we need to instantiate an evar with the output of the VM.

**Comment by JasonGross** [comment]
> 4\. fiat-crypto bootstrapping, [IdentifiersGENERATEDProofs](https://coq.gitlab.io/-/coq/-/jobs/3375061712/artifacts/_bench/html/coq-fiat-crypto-with-bedrock/src/Language/IdentifiersGENERATEDProofs.v.html) Not sure about this one. 1.5s slowdown before Qed and then Qed with barely any slowdown.

Presumably this is
https://github.com/mit-plv/rewriter/blob/de56173b42be33ea3bcf4f96a491bd5c2b30892e/src/Rewriter/Language/IdentifiersGenerateProofs.v#L114
where we `try (vm_compute; reflexivity)` on something like a hundred goals, one for each identifier.

**Comment by ppedrot** [comment]
Closing this for now. We need a PR that retypechecks the full rel+named context.

**Pull request closed without merging**